src/fvectorbin/nnet3-fvector-perturb-signal.cc

```diff
@@ -36,6 +36,78 @@ struct NoiseController{
     noise_t_start_(noise_t_start), noise_t_end_(noise_t_end), snr_(snr) { }
 };
 
+void GenerateController(std::vector<std::string> &segments, 
+                        std::vector<NoiseController> *controller) {
+  BaseFloat wav_t_start;
+  BaseFloat wav_t_end;
+  std::string noise_uttid;
+  BaseFloat noise_t_start;
+  BaseFloat noise_t_end;
+  BaseFloat snr;
+  for(int i=0; i < segments.size(); ++i) {
+    std::vector<std::string> split_string;
+    SplitStringToVector(segments[i], ":", true, &split_string);
+    KALDI_ASSERT(split_string.size() == 6);
+    ConvertStringToReal(split_string[0], &wav_t_start);
+    ConvertStringToReal(split_string[1], &wav_t_end);
+    noise_uttid = split_string[2];
+    ConvertStringToReal(split_string[3], &noise_t_start);
+    ConvertStringToReal(split_string[4], &noise_t_end);
+    ConvertStringToReal(split_string[5], &snr);
+  
+    controller->push_back(NoiseController(wav_t_start, wav_t_end, noise_uttid,
+                                            noise_t_start, noise_t_end, snr));
+  }
+}
+
+void ApplyNoise(std::string &noise_scp, const std::vector<NoiseController> &controller,
+                const VectorBase<BaseFloat> &input_wav, VectorBase<BaseFloat> *perturbed_wav) {
+  // about noise list
+  RandomAccessTableReader<WaveHolder> noise_reader(noise_scp);
+  int samp_freq_input = input_wav.Dim();
+
+  // add noise
+
+  for (int i=0; i < controller.size(); ++i) {
+    const WaveData &noise_wav = noise_reader.Value(controller[i].noise_uttid_);
```

Comment:
variable in struct should not have "_" at end, just class variable has "_". Also it should be noise_id not noise_uttid!!